src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -190,25 +196,29 @@ private Task StartAsyncCore()
                     await _eventQueue.Drain();
 
                     await Task.WhenAny(_eventQueue.Drain().NoThrow(), Task.Delay(_eventQueueDrainTimeout));
-                    _httpClient?.Dispose();
 
                     _logger.CompleteClosed(_connectionId);
+
+                    // At this point the connection can be either in the Connected or Disposed state. The state should be changed
+                    // to the Disconnected state only if it was in the Connected state.
+                    ChangeState(from: ConnectionState.Connected, to: ConnectionState.Disconnected);
+
+                    _closeTcs.SetResult(null);
+
                     if (t.IsFaulted)
                     {
-                        _closedTcs.TrySetException(t.Exception.InnerException);
+                        Closed?.Invoke(t.Exception.InnerException);
```

Comment:
We need to `try catch` around these and log `ExceptionThrownFromCallback`

Reply:
No. Exceptions should not be thrown from event handlers.